pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerBase.java

```diff
@@ -148,25 +139,53 @@ protected ConsumerBase(PulsarClientImpl client, String topic, ConsumerConfigurat
                     "Cannot use receive() when a listener has been set");
         }
 
-        switch (getState()) {
-        case Ready:
-        case Connecting:
-            break; // Ok
-        case Closing:
-        case Closed:
-            throw new PulsarClientException.AlreadyClosedException("Consumer already closed");
-        case Terminated:
-            throw new PulsarClientException.AlreadyClosedException("Topic was terminated");
-        case Failed:
-        case Uninitialized:
-            throw new PulsarClientException.NotConnectedException();
+        PulsarClientException exception = verifyConsumerState();
+        if (exception != null) {
+            throw exception;
         }
-
         return internalReceive(timeout, unit);
     }
 
     abstract protected Message<T> internalReceive(int timeout, TimeUnit unit) throws PulsarClientException;
 
+    @Override
+    public Messages<T> batchReceive() throws PulsarClientException {
+        if (listener != null) {
```

Comment:
the batchReceive should just call `batchReceiveAsync` to avoid duplicated code

Reply:
I have added a new method for verify batch receive, because batchReceive() need to handle  InterruptedException and ExecutionException and then unwrap them.